File: gingkocc/gingko.cpp
///////////////////////////////////////////////////////////////////////////////
//
// GINGKO Biogeographical Evolution Simulator.
//
// Copyright 2009 Jeet Sukumaran and Mark T. Holder.
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License along
// with this program. If not, see <http://www.gnu.org/licenses/>.
//
///////////////////////////////////////////////////////////////////////////////

#include "gingko_defs.hpp"
#include "confsys.hpp"
#include "biosys.hpp"
#include "world.hpp"
#include "tree.hpp"
#include "cmdopt.hpp"
#include "filesys.hpp"
#include <iostream>
#include <cstdlib>
#include <vector>
#include <set>
#include <ctime>
#include <string>
#include <sstream>


int main(int argc, char* argv[]) {

    std::string config_filepath;
    std::string output_dir = ".";
    std::string replicate_id;
    bool validate_config_only = false;
    unsigned long rand_seed = 0;

    gingko::OptionParser parser = gingko::OptionParser("Gingko 0.1.3",
            "Gingko Biogeographical Evolution Simulator",
            "%prog [options] <CONFIGURATION-FILEPATH>");
    
    parser.add_option<std::string>(&replicate_id, "-i", "--replicate-id", 
                               "optional identifier to add to names of output files produced during this run");
    parser.add_option<std::string>(&output_dir, "-o", "--output-dir", 
                                   "directory to which to save output files (default = current)");
    parser.add_switch(&validate_config_only, "-v", "--validate",
                      "load and process configuration file to check for errors, but do not actually execute run");
    parser.add_option<unsigned long>(&rand_seed, "-z", "--random-seed", 
                               "random seed for this run (overrides seed set in configuration file)");

    parser.parse(argc, argv);       
    
    std::vector<std::string> args = parser.get_args();
    
    if (args.size() == 0) {
        parser.write_usage(std::cout);
        exit(1);
    }
    
    gingko::World world;
    world.set_replicate_id(replicate_id);
    world.set_output_dir(output_dir);
       
    gingko::confsys::configure_world(world, args[0]);
    if (validate_config_only) {
        std::cout << "World configured using: \"" + args[0] + "\"" << std::endl; 
        std::cout << "Configuration file validates." << std:: endl;
    } else {
        world.open_logs();
        world.log_info("World configured using: \"" + args[0] + "\"");
        
        std::ostringstream seed_info;
        if (parser.is_set("--random-seed")) {
            seed_info << "Setting random seed from command line: " << rand_seed << ".";
            world.set_random_seed(rand_seed);
        } else {
            seed_info << "Using random seed: " << world.get_random_seed() << "."; 
        }
        world.log_info(seed_info.str());
        world.run();
        world.close_logs();
    }
    
}

File: gingkopy/gingkopy/gingko.py
#! /usr/bin/env python

###############################################################################
##
## GINGKO Biogeographical Evolution Simulator Post-Processing Library.
##
## Copyright 2009 Jeet Sukumaran and Mark T. Holder.
##
## This program is free software; you can redistribute it and#or modify
## it under the terms of the GNU General Public License as published by
## the Free Software Foundation; either version 3 of the License, or
## (at your option) any later version.
## 
## This program is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU General Public License for more details.
## 
## You should have received a copy of the GNU General Public License along
## with this program. If not, see <http:##www.gnu.org#licenses#>.
##
###############################################################################

import re
import math

try:
    from rpy2 import robjects
except:
    pass

from dendropy import splits
from dendropy import datasets
from dendropy import treedists

def read_trees(src, encode_splits=True, dataset=None):
    """
    Wraps reading of trees, georeferencing taxa, and encoding of splits
    """
    if dataset is None:
        dataset = datasets.Dataset()
    if isinstance(src, str):
        f = open(src, "rU")
    else:
        f = src
    gtrees = dataset.read_trees(f, "NEXUS")
    georeference_taxa(dataset.taxa_blocks[0])
    if encode_splits:
        for t in gtrees:
            splits.encode_splits(t)
    return gtrees            
    
def georeference_taxa(taxa_block):
    """
    Decorates taxa with x and y cell coordinates as parsed from the taxon 
    labels.
    """
    for taxon in taxa_block:
        taxon.x = int(re.match(".* x([\d]+) .*", taxon.label).groups(1)[0])
        taxon.y = int(re.match(".* y([\d]+) .*", taxon.label).groups(1)[0])        
    
def euclidean_distance(p1, p2):
    """
    Returns the euclidean distance between two points, `p1` and `p2`, which
    each is tuple of two integers, (x,y).
    """
    return math.sqrt( pow(p1[0] - p2[0], 2) + pow(p1[1] - p2[1], 2) )
            
def geo_distance(node1, node2):
    """
    Returns the spatial euclidean distance between two nodes, assuming that the
    node taxa have been marked up with "x" and "y" attributes.
    """
    return euclidean_distance((node1.taxon.x, node1.taxon.y), (node2.taxon.x, node2.taxon.y))

def calc_pearsons_r(tree):
    """
    Calculates Pearson's R on a georeferenced tree for geographic vs. patristic 
    distances. Taxa on trees must be georeferenced, and splits encoded.
    """
    gd = []
    pd = []
    leaves = [x for x in tree.leaf_iter()]
    for idx1, leaf1 in enumerate(leaves):
        for idx2, leaf2 in enumerate(leaves[idx1+1:]):
            gd.append(geo_distance(leaf1, leaf2))
            pd.append(treedists.patristic_distance(tree, leaf1.taxon, leaf2.taxon))
    k = robjects.r.cor(robjects.IntVector(gd), robjects.IntVector(pd), method='pearson')
    tree.pearson_r_val = k[0]
    return tree.pearson_r_val
    